runtime/near-vm-logic/src/mocks/mock_external.rs

```diff
@@ -220,6 +233,136 @@ impl External for MockedExternal {
+            } else {
+                let mut acc = G1::zero();
+                for i in 0..size {
+                    let chunk = &data[97 * i..97 * (i + 1)];
+                    let neg_sign = match chunk[0] {
+                        0u8 => Ok(false),
+                        1u8 => Ok(true),
+                        _ => Err(VMLogicError::ExternalError(b"Wrong sign boolean value".to_vec())),
+                    }?;
+                    let e = parse_fr_le(&chunk[1..33])
+                        .ok_or(VMLogicError::ExternalError(b"Invalid exponent".to_vec()))?;
+                    let b_x = parse_fq_le(&chunk[33..65])
+                        .ok_or(VMLogicError::ExternalError(b"Invalid x coordinate".to_vec()))?;
+                    let b_y = parse_fq_le(&chunk[65..97])
+                        .ok_or(VMLogicError::ExternalError(b"Invalid y coordinate".to_vec()))?;
+
+                    if b_x.is_zero() && b_y.is_zero() {
+                        continue;
+                    }
+
+                    let b = G1::from(AffineG1::new(b_x, b_y).map_err(|_| {
+                        VMLogicError::ExternalError(b"Invalid b argument - not on curve".to_vec())
+                    })?);
+
+                    if e.is_zero() {
+                        continue;
+                    }
+
+                    acc = if neg_sign { acc - b * e } else { acc + b * e };
+                }
+
+                if acc.is_zero() {
+                    Ok(vec![0u8; 64])
+                } else {
+                    let p = AffineG1::from_jacobian(acc)
+                        .ok_or(VMLogicError::ExternalError(b"Wrong acc point data".to_vec()))?;
+                    let mut x = [0u8; 32];
+                    p.x().to_big_endian(&mut x[..]).unwrap();
+                    let mut y = [0u8; 32];
+                    p.y().to_big_endian(&mut y[..]).unwrap();
+                    Ok(x.iter().rev().chain(y.iter().rev()).cloned().collect::<Vec<u8>>())
+                }
+            }
+        }
+    }
+
+    fn alt_bn128_pairing_check(&self, data: &[u8]) -> Result<bool> {
+        if data.len() < 4 {
+            Err(VMLogicError::ExternalError(b"Wrong argument length".to_vec()))
+        } else {
+            let size = u32::from_le_bytes(data[0..4].try_into().unwrap()) as usize;
+            let data = &data[4..];
+
+            if data.len() != 192 * size {
+                Err(VMLogicError::ExternalError(b"Wrong argument length".to_vec()))
+            } else if data.is_empty() {
+                Ok(true)
+            } else {
+                // (a, b_a, b_b - each 64-byte affine coordinates)
+                let elements = data.len() / 192;
+                let mut vals = Vec::new();
+                for idx in 0..elements {
+                    let a_x = parse_fq_le(&data[idx * 192..idx * 192 + 32]).ok_or(
+                        VMLogicError::ExternalError(b"Invalid a argument x coordinate".to_vec()),
```

Comment:
We cannot throw `ExternalError` here. `ExternalError` is only for errors like out of memory, corrupted database, etc. When it is thrown the node shuts down. The contract should not be able to cause `ExternalError` by providing some specific input. Instead we should define a set of Errors specific to alt_bn128 math and throw them here.

Reply:
Now HostError::GuestPanic is used.